### PR TITLE
[query] sort, topk, bottomk fixes for instant queries

### PR DESCRIPTION
**Pull request by soundvibe**
**What this PR does / why we need it**:

This PR follows the work started with https://github.com/m3db/m3/pull/2756.
It implements sort functionality and  additionally fixes the issues when NaN values are not returned for instant queries in some cases. 
It would be enough to merge only this PR because it contains all the changes from https://github.com/m3db/m3/pull/2756.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
topk and bottomk results for instant queries are the same except that they are now sorted.
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

File: src/query/parser/promql/matchers.go [hunk 1]
```diff
@@ -262,10 +262,11 @@ func NewFunctionExpr(
 		p, err = scalar.NewTimeOp(tagOptions)
 		return p, true, err
 
-	// NB: no-ops.
 	case linear.SortType, linear.SortDescType:
-		return nil, false, err
+		p, err = linear.NewSortOp(name)
```
Comment (arnikola) [inline src/query/parser/promql/matchers.go:266]: :+1:

File: src/query/models/query_context.go [hunk 2]
```diff
@@ -48,6 +48,8 @@ type QueryContextOptions struct {
 	LimitMaxDocs int
 	// RequireExhaustive results in an error if the query exceeds the series limit.
 	RequireExhaustive bool
+	// Is instant query
+	Instantaneous     bool
 	RestrictFetchType *RestrictFetchTypeQueryContextOptions
```
Comment (arnikola) [inline src/query/models/query_context.go:53]: nit: make these godoc style comments;
```suggestion
	// Instantaneous indicates an instant query.
	Instantaneous     bool
	// RestrictFetchType restricts the query fetches.
	RestrictFetchType *RestrictFetchTypeQueryContextOptions
```

File: src/query/block/meta.go [hunk 3]
```diff
@@ -55,6 +55,15 @@ func (m Metadata) String() string {
...
 type Warnings []Warning
 
+// Nullable bool
+type Bool int
+
+const (
+	BoolNull = iota
+	BoolFalse
+	BoolTrue
+)
```
Comment (arnikola) [inline src/query/block/meta.go:65]: No 3 value bools please

File: src/query/api/v1/handler/prometheus/native/read.go [hunk 4]
```diff
@@ -135,8 +136,16 @@ func (h *promReadHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	handleroptions.AddWarningHeaders(w, result.Meta)
 	h.promReadMetrics.fetchSuccess.Inc(1)
 
+	keepNans := h.opts.Config().ResultOptions.KeepNans
+	switch result.Meta.KeepNans {
+		case block.BoolFalse:
+			keepNans = false
+		case block.BoolTrue:
+			keepNans = true
+	}
```
Comment (arnikola) [inline src/query/api/v1/handler/prometheus/native/read.go:145]: Don't think we'd ever want to overwrite the case where the config keeps Nans, so this can be simplified after making `result.Meta.KeepNans` an actual bool

```suggestion
	keepNans := h.opts.Config().ResultOptions.KeepNans
	if !keepNans {
		keepNans = result.Meta.KeepNans 
	}
```

Also, can we rename `Nans` to `NaNs` within codebase? Probably fine to leave that to another PR though

File: src/query/block/meta.go [hunk 5]
```diff
@@ -68,6 +77,8 @@ type ResultMetadata struct {
 	Warnings Warnings
 	// Resolutions is a list of resolutions for series obtained by this query.
 	Resolutions []time.Duration
+	// KeepNans indicates if NaNs should be kept when returning results
```
Comment (arnikola) [inline src/query/block/meta.go:80]: nti: `.` at end of comment

File: src/query/functions/aggregation/take_test.go [hunk 6]
```diff
@@ -111,9 +172,9 @@ func TestTakeTopFunctionFilteringWithoutA(t *testing.T) {
 	}
 
 	// Should have the same metas as when started
-	assert.Equal(t, seriesMetas, sink.Metas)
+	require.Equal(t, seriesMetas, sink.Metas)
```
Comment (arnikola) [inline src/query/functions/aggregation/take_test.go:175]: Any reason for switching these from assert to require?

File: src/query/functions/aggregation/take_test.go [hunk 7]
```diff
@@ -59,6 +105,20 @@ func TestTakeFn(t *testing.T) {
...
+		{Tags: test.StringTagsToTags(test.StringTags{{N: "job", V: "api-server"}, {N: "instance", V: "1"}, {N: "group", V: "production"}})},
+		{Tags: test.StringTagsToTags(test.StringTags{{N: "job", V: "api-server"}, {N: "instance", V: "2"}, {N: "group", V: "production"}})},
+		{Tags: test.StringTagsToTags(test.StringTags{{N: "job", V: "api-server"}, {N: "instance", V: "0"}, {N: "group", V: "canary"}})},
+		{Tags: test.StringTagsToTags(test.StringTags{{N: "job", V: "api-server"}, {N: "instance", V: "1"}, {N: "group", V: "canary"}})},
+		{Tags: test.StringTagsToTags(test.StringTags{{N: "job", V: "app-server"}, {N: "instance", V: "0"}, {N: "group", V: "production"}})},
+		{Tags: test.StringTagsToTags(test.StringTags{{N: "job", V: "app-server"}, {N: "instance", V: "1"}, {N: "group", V: "production"}})},
+		{Tags: test.StringTagsToTags(test.StringTags{{N: "job", V: "app-server"}, {N: "instance", V: "0"}, {N: "group", V: "canary"}})},
+		{Tags: test.StringTagsToTags(test.StringTags{{N: "job", V: "app-server"}, {N: "instance", V: "1"}, {N: "group", V: "canary"}})},
+	}
+)
```
Comment (arnikola) [inline src/query/functions/aggregation/take_test.go:120]: nit: define this above where it's used; in fact it's only used in `TestTakeInstantFn`  so better to just add it within the function

File: src/query/functions/linear/sort.go [hunk 8]
```diff
@@ -20,13 +20,178 @@
...
+	}
+
+	instantaneous := queryCtx.Options.Instantaneous
+	meta := b.Meta()
+	seriesMetas := utils.FlattenMetadata(meta, stepIter.SeriesMeta())
+
+	if instantaneous {
+		return n.processInstantBlock(queryCtx, stepIter, meta, seriesMetas)
+	}
+	return n.processNopBlock(queryCtx, stepIter, meta, seriesMetas)
```
Comment (arnikola) [inline src/query/functions/linear/sort.go:101]: Since this no-ops on vectors, can just return block

```suggestion
	if !queryCtx.Options.Instantaneous {
		return block, nil
	}

	stepIter, err := b.StepIter()
	if err != nil {
		return nil, err
	}
 
	meta := b.Meta()
	seriesMetas := utils.FlattenMetadata(meta, stepIter.SeriesMeta()) 
	return n.processInstantBlock(queryCtx, stepIter, meta, seriesMetas)  // or just have the body of processInstantBlock here
```

File: src/query/functions/linear/sort.go [hunk 9]
```diff
@@ -20,13 +20,178 @@
...
+}
+
+func (n *sortNode) processInstantBlock(queryCtx *models.QueryContext, stepIter block.StepIter, meta block.Metadata, seriesMetas []block.SeriesMeta) (block.Block, error) {
+	for index := 0; stepIter.Next(); index++ {
+		if isLastStep(index, stepIter.StepCount()) {
+			meta.ResultMetadata.KeepNans = block.BoolTrue
+			//we only care for the last step values for the instant query
+			values := stepIter.Current().Values()
+			valuesToSort := make([]valueAndMeta, len(values))
```
Comment (arnikola) [inline src/query/functions/linear/sort.go:110]: nit: using append is generally safer than indexed inserts, and doesn't make any real difference in performance

File: src/query/functions/linear/sort.go [hunk 10]
```diff
@@ -20,13 +20,178 @@
...
+
+	if instantaneous {
+		return n.processInstantBlock(queryCtx, stepIter, meta, seriesMetas)
+	}
+	return n.processNopBlock(queryCtx, stepIter, meta, seriesMetas)
+}
+
+func (n *sortNode) processInstantBlock(queryCtx *models.QueryContext, stepIter block.StepIter, meta block.Metadata, seriesMetas []block.SeriesMeta) (block.Block, error) {
+	for index := 0; stepIter.Next(); index++ {
+		if isLastStep(index, stepIter.StepCount()) {
```
Comment (arnikola) [inline src/query/functions/linear/sort.go:106]:
```golang
stepsToSkip := stepIter.StepCount() - 1
for i := 0; i < stepsToSkip; i++ {
	if !stepIter.Next() {
		return nil, fmt.Errorf("invalid step count; expected %d got %d", stepsToSkip+1, i)
	}
}
```

File: src/query/functions/linear/sort.go [hunk 11]
```diff
@@ -20,13 +20,178 @@
...
+		if err := builder.AppendValues(index, stepIter.Current().Values()); err != nil {
+			return nil, err
+		}
+	}
+
+	if err = stepIter.Err(); err != nil {
+		return nil, err
+	}
+	return builder.Build(), nil
+}
```
Comment (arnikola) [inline src/query/functions/linear/sort.go:177]: nit; can remove this

File: src/query/functions/linear/sort.go [hunk 12]
```diff
@@ -20,13 +20,178 @@
...
+
+func isLastStep(stepIndex int, stepCount int) bool {
+	return stepIndex == stepCount-1
+}
```
Comment (arnikola) [inline src/query/functions/linear/sort.go:181]: nit: can remove this

File: src/query/functions/linear/sort_test.go [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestShouldFailWhenOpTypeIsInvalid(t *testing.T) {
+	_, err := NewSortOp("sortAsc")
+	require.Error(t, err)
+}
+
+func TestAscSort(t *testing.T) {
```
Comment (arnikola) [inline src/query/functions/linear/sort_test.go:19]: These should probably be testing the entire `NewSortOp` flow instead

File: src/query/functions/utils/heap.go [hunk 14]
```diff
@@ -82,7 +99,7 @@ func NewFloatHeap(isMaxHeap bool, capacity int) FloatHeap {
 }
 
 // Push pushes a value and index pair to the heap
-func (fh FloatHeap) Push(value float64, index int) {
+func (fh *FloatHeap) Push(value float64, index int) {
```
Comment (arnikola) [inline src/query/functions/utils/heap.go:102]: Do these need to be pointers?

File: src/query/functions/utils/heap.go [hunk 15]
```diff
@@ -36,14 +38,29 @@ func maxHeapLess(i, j ValueIndexPair) bool {
...
+
+func DescFloat64(i, j float64) bool {
+	return i > j || math.IsNaN(j) && !math.IsNaN(i)
+}
+
+func Min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
```
Comment (arnikola) [inline src/query/functions/utils/heap.go:63]: nit: This doesn't really need to be a function

File: src/query/functions/utils/heap.go [hunk 16]
```diff
@@ -36,14 +38,29 @@ func maxHeapLess(i, j ValueIndexPair) bool {
...
+	return i.Val > j.Val || math.IsNaN(i.Val) && !math.IsNaN(j.Val)
+}
+
+func AscFloat64(i, j float64) bool {
+	return i < j || math.IsNaN(j) && !math.IsNaN(i)
+}
+
+func DescFloat64(i, j float64) bool {
+	return i > j || math.IsNaN(j) && !math.IsNaN(i)
+}
```
Comment (arnikola) [inline src/query/functions/utils/heap.go:57]: nit: these should have more descriptive names like `GreaterWithNaNs`, and will need comments

File: src/query/functions/utils/heap.go [hunk 17]
```diff
@@ -36,14 +38,29 @@ func maxHeapLess(i, j ValueIndexPair) bool {
 	if i.Val == j.Val {
 		return i.Index > j.Index
 	}
-	return i.Val < j.Val
+	return math.IsNaN(i.Val) && !math.IsNaN(j.Val) || i.Val < j.Val
```
Comment (arnikola) [inline src/query/functions/utils/heap.go:41]: This will need to sort by `Index` if both `i` and `j` are NaN

File: src/query/functions/utils/heap.go [hunk 18]
```diff
@@ -36,14 +38,29 @@ func maxHeapLess(i, j ValueIndexPair) bool {
...
 }
 
 func minHeapLess(i, j ValueIndexPair) bool {
 	if i.Val == j.Val {
 		return i.Index > j.Index
 	}
-	return i.Val > j.Val
+	return i.Val > j.Val || math.IsNaN(i.Val) && !math.IsNaN(j.Val)
```
Comment (arnikola) [inline src/query/functions/utils/heap.go:48]: This will need to sort by `Index` if both `i` and `j` are NaN

File: src/query/functions/aggregation/take.go [hunk 19]
```diff
@@ -39,39 +39,50 @@ const (
...
 	k := int(params.Parameter)
+
 	if k < 1 {
-		fn = func(values []float64, buckets [][]int) []float64 {
-			return takeNone(values, buckets)
+		fn = func(heap utils.FloatHeap, values []float64, buckets [][]int) []float64 {
+			return takeNone(values)
+		}
+		fnInstant = func(heap utils.FloatHeap, values []float64, buckets [][]int, seriesMetas []block.SeriesMeta) []valueAndMeta {
+			return takeInstantNone(values, seriesMetas)
```
Comment (arnikola) [inline src/query/functions/aggregation/take.go:65]: Can we just get rid of this `k < 1` special casing? It's a silly optimization which adds minimal performance for the additional complexity

File: src/query/functions/aggregation/take.go [hunk 20]
```diff
@@ -39,39 +39,50 @@ const (
...
 }
 
 // takeOp stores required properties for take ops
 type takeOp struct {
-	params   NodeParams
-	opType   string
-	takeFunc takeFunc
+	params          NodeParams
+	opType          string
+	k				int
```
Comment (arnikola) [inline src/query/functions/aggregation/take.go:83]: nit: spacing

File: src/query/functions/aggregation/take.go [hunk 21]
```diff
@@ -95,11 +106,13 @@ func (o takeOp) Node(
...
 
-func newTakeOp(params NodeParams, opType string, takeFunc takeFunc) takeOp {
+func newTakeOp(params NodeParams, opType string, k int, takeFunc takeFunc, takeInstantFunc takeInstantFunc) takeOp {
 	return takeOp{
-		params:   params,
-		opType:   opType,
-		takeFunc: takeFunc,
+		params:          params,
+		opType:          opType,
+		k: 				 k,
```
Comment (arnikola) [inline src/query/functions/aggregation/take.go:113]: nit; spacing

File: src/query/functions/aggregation/take.go [hunk 22]
```diff
@@ -136,45 +155,134 @@ func (n *takeNode) ProcessBlock(queryCtx *models.QueryContext, ID parser.NodeID,
...
 
-	// retain original metadatas
-	builder, err := n.controller.BlockBuilder(queryCtx, meta, stepIter.SeriesMeta())
-	if err != nil {
-		return nil, err
+	seriesCount := maxSeriesCount(buckets)
+	if instantaneous {
+		heap := utils.NewFloatHeap(takeTop, utils.Min(n.op.k, seriesCount))
+		return n.processBlockInstantaneous(heap, queryCtx, meta, stepIter, seriesMetas, buckets)
+	} else {
```
Comment (arnikola) [inline src/query/functions/aggregation/take.go:162]: nit; this doesn't need an else

File: src/query/functions/aggregation/take.go [hunk 23]
```diff
@@ -136,45 +155,134 @@ func (n *takeNode) ProcessBlock(queryCtx *models.QueryContext, ID parser.NodeID,
...
+func (n *takeNode) processBlockInstantaneous(
+		heap utils.FloatHeap,
+		queryCtx *models.QueryContext,
+		metadata block.Metadata,
+		stepIter block.StepIter,
+		seriesMetas []block.SeriesMeta,
+		buckets [][]int) (block.Block, error) {
+	stepCount := stepIter.StepCount()
+	metadata.ResultMetadata.KeepNans = block.BoolTrue
+	for index := 0; stepIter.Next(); index++ {
```
Comment (arnikola) [inline src/query/functions/aggregation/take.go:220]: nit: same changes here as suggested in sort.go

File: src/query/functions/aggregation/take.go [hunk 24]
```diff
@@ -136,45 +155,134 @@ func (n *takeNode) ProcessBlock(queryCtx *models.QueryContext, ID parser.NodeID,
...
+			if err = blockBuilder.AddCols(1); err != nil {
+				return nil, err
+			}
+			if err := blockBuilder.AppendValues(0, blockValues); err != nil {
+				return nil, err
+			}
+			return blockBuilder.Build(), nil
+		}
 	}
+	return nil, fmt.Errorf("no data to return: %s", n.op.opType)
```
Comment (arnikola) [inline src/query/functions/aggregation/take.go:257]: This will need to check `StepIter.Err()` too

File: src/query/functions/aggregation/take_test.go (hunk 6 above)
Comment (soundvibe) [reply src/query/functions/aggregation/take_test.go:175]: These tests were already using `require` so my intention was to unify assertions. I can for sure replace some checks to use `assert` so that failing tests wont stop further test execution.

File: src/query/functions/linear/sort.go (hunk 8 above)
Comment (soundvibe) [reply src/query/functions/linear/sort.go:101]: Nice! Didn't thought that it's the same block ;)

File: src/query/functions/linear/sort.go (hunk 9 above)
Comment (soundvibe) [reply src/query/functions/linear/sort.go:110]: Not sure how current code is unsafe? `append` needs to check current capacity and grow when needed which we do not want here (more possible allocations) because we already know the actual length and capacity for our value arrays.

File: src/query/functions/utils/heap.go (hunk 14 above)
Comment (soundvibe) [reply src/query/functions/utils/heap.go:102]: I think it's more clear this way because most of our functions mutates internal state. Also, generally it is advised to use either pointer receivers or value receivers on a type, not a mixture of both.

File: src/query/functions/aggregation/take.go (hunk 19 above)
Comment (soundvibe) [reply src/query/functions/aggregation/take.go:65]: Completely agree.

File: src/query/functions/utils/heap.go (hunk 15 above)
Comment (soundvibe) [reply src/query/functions/utils/heap.go:65]: What could you suggest instead?
To me, the code is more readable this way.

File: src/query/functions/linear/sort_test.go [hunk 25]
```diff
@@ -0,0 +1,121 @@
+package linear
```
Comment (arnikola) [inline src/query/functions/linear/sort_test.go:1]: This needs a header

File: src/query/functions/linear/sort_test.go [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+func processSortOp(t *testing.T, op parser.Params, instant bool) *executor.SinkNode {
+	bl := test.NewBlockFromValuesWithSeriesMeta(bounds, seriesMetas, v)
+	c, sink := executor.NewControllerWithSink(parser.NodeID(1))
+	node := op.(sortOp).Node(c, transform.Options{})
+	queryContext := models.NoopQueryContext()
+	queryContext.Options.Instantaneous = instant
+	err := node.Process(queryContext, parser.NodeID(0), bl)
+	require.NoError(t, err)
+	return sink
+}
```
Comment (arnikola) [inline src/query/functions/linear/sort_test.go:121]: nit; newline

File: src/query/functions/utils/heap.go (hunk 14 above)
Comment (arnikola) [reply src/query/functions/utils/heap.go:104]: None of these mutate the floatheap itself, just the internal data structure; these can all be reverted

File: src/query/functions/aggregation/take.go [hunk 27]
```diff
@@ -136,8 +141,18 @@ func (n *takeNode) ProcessBlock(queryCtx *models.QueryContext, ID parser.NodeID,
 		seriesMetas,
 	)
 
-	// retain original metadatas
-	builder, err := n.controller.BlockBuilder(queryCtx, meta, stepIter.SeriesMeta())
+	seriesCount := maxSeriesCount(buckets)
+	if instantaneous {
+		heap := utils.NewFloatHeap(takeTop, utils.Min(n.op.k, seriesCount))
```
Comment (arnikola) [inline src/query/functions/aggregation/take.go:146]:
```suggestion
		heapSize := seriesCount
		if n.op.k < seriesCount {
			heapSize = n.op.k
		}
		
		heap := utils.NewFloatHeap(takeTop, heapSize)
```

File: src/query/functions/utils/heap_test.go [hunk 28]
```diff
@@ -216,3 +220,133 @@ func TestNegativeCapacityHeap(t *testing.T) {
 		assert.Equal(t, testArray[pair.Index], pair.Val)
 	}
 }
+
+func equalPairs(t *testing.T, expected, actual []ValueIndexPair) {
+	e := fmt.Sprint(expected)
+	a := fmt.Sprint(actual)
+	assert.Equal(t, e, a)
```
Comment (arnikola) [inline src/query/functions/utils/heap_test.go:227]: Matching strings here seems odd; can this not use `EqualsWithNaNs`?

File: src/query/functions/utils/heap.go [hunk 29]
```diff
@@ -33,17 +35,34 @@ type ValueIndexPair struct {
 type lessFn func(ValueIndexPair, ValueIndexPair) bool
 
 func maxHeapLess(i, j ValueIndexPair) bool {
-	if i.Val == j.Val {
+	if i.Val == j.Val || math.IsNaN(i.Val) && math.IsNaN(j.Val) {
```
Comment (arnikola) [inline src/query/functions/utils/heap.go:38]: nit: These conditionals are very hard to read and understand, can you make them more clear and verbose?

```suggestion
func maxHeapLess(i, j ValueIndexPair) bool {	func maxHeapLess(i, j ValueIndexPair) bool {
	if math.IsNaN(i.Val) {
		if math.IsNaN(j.Val) {
			return i.Index > j.Index
		}
		
		return true
	}
	
	if math.IsNaN(j.Val) {
		return true
	}
	
	if i.Val == j.Val {
		return i.Index > j.Index
	}
	
	return i.Val < j.Val
}
```

File: src/query/functions/utils/heap_test.go (hunk 28 above)
Comment (soundvibe) [reply src/query/functions/utils/heap_test.go:227]: It doesn't work when using `EqualsWithNaNs`:
```
panic: interface conversion: interface {} is []utils.ValueIndexPair, not string [recovered]
	panic: interface conversion: interface {} is []utils.ValueIndexPair, not string
```

File: src/query/functions/aggregation/take.go [hunk 30]
```diff
@@ -198,3 +274,28 @@ func takeFn(heap utils.FloatHeap, values []float64, buckets [][]int) []float64 {
 
 	return values
 }
+
+func takeInstantFn(heap utils.FloatHeap, values []float64, buckets [][]int, metas []block.SeriesMeta) []valueAndMeta {
+	var result []valueAndMeta
```
Comment (arnikola) [inline src/query/functions/aggregation/take.go:279]: nit: init with size `result := make([]valueAndMeta, 0, len(values))`

File: src/query/functions/aggregation/take.go [hunk 31]
```diff
@@ -198,3 +274,28 @@ func takeFn(heap utils.FloatHeap, values []float64, buckets [][]int) []float64 {
...
+	if heap.Cap() < 1 {
+		return result
```
Comment (arnikola) [inline src/query/functions/aggregation/take.go:281]: This may need to nil out values; `util.Memset(values, math.NaN())`

File: src/query/functions/aggregation/take.go (hunk 27 above)
Comment (soundvibe) [reply src/query/functions/aggregation/take.go:146]: Sorry, but I don't understand your point here. Why do you think this is better? IMO, you've just added more code and reduced readability. Now I need to read 5 more lines of code in order to understand the logic how `heapSize` is calculated. Also, by doing this you essentially made this logic more difficult to test...

File: src/query/functions/utils/heap_test.go (hunk 28 above)
Comment (arnikola) [reply src/query/functions/utils/heap_test.go:227]: You would need to either extend EqualsWithNaNs to take this type, or better to compare the values by hand

File: src/query/functions/linear/sort.go [hunk 32]
```diff
@@ -20,13 +20,148 @@
...
+	}
+	values := stepIter.Current().Values()
+	meta.ResultMetadata.KeepNaNs = true
+	valuesToSort := make([]valueAndMeta, len(values))
+	for i, value := range values {
+		valuesToSort[i] = valueAndMeta{
+			val:        value,
+			seriesMeta: seriesMetas[i],
+		}
+	}
```
Comment (arnikola) [inline src/query/functions/linear/sort.go:117]: nit: safer to use append here

File: src/query/functions/aggregation/take.go [hunk 33]
```diff
@@ -147,34 +162,95 @@ func (n *takeNode) ProcessBlock(queryCtx *models.QueryContext, ID parser.NodeID,
...
+	return blockBuilder.Build(), nil
+}
+
+func mapToValuesAndSeriesMetas(takenSortedValues []valueAndMeta) ([]float64, []block.SeriesMeta) {
+	blockValues := make([]float64, len(takenSortedValues))
+	blockSeries := make([]block.SeriesMeta, len(takenSortedValues))
+	for i, sortedValue := range takenSortedValues {
+		blockValues[i] = sortedValue.val
+		blockSeries[i] = sortedValue.seriesMeta
+	}
```
Comment (arnikola) [inline src/query/functions/aggregation/take.go:239]: Safer to use append here

File: src/query/functions/aggregation/take.go (hunk 31 above)
Comment (soundvibe) [reply src/query/functions/aggregation/take.go:281]: There is no need for this because we are not mutating `values` in this method or using them afterwards.

File: src/query/functions/utils/heap_test.go (hunk 28 above)
Comment (soundvibe) [reply src/query/functions/utils/heap_test.go]: https://github.com/m3db/m3/pull/2792/commits/bd3f326bd21383bca9c81fc1f51927a935bf619f#diff-cf04475fadbfc80ae239ceb87d393ce927af7b017dd971ad6b741cee31e15dcbR224-R228

File: src/query/functions/utils/heap.go (hunk 14 above)
Comment (soundvibe) [reply src/query/functions/utils/heap.go:104]: As I was describing before, I've done this for better clarity and intent. 
Also, I am not the only one thinking like this: https://github.com/golang/go/wiki/CodeReviewComments#receiver-type 

Here it states exactly the same situation we have here:
> If the receiver is a struct, array or slice and any of its elements is a pointer to something that might be mutating, prefer a pointer receiver, as it will make the intention more clear to the reader.

I can for sure revert it to be non pointer, but why this would better and what are the downsides of the current approach?

File: src/query/functions/utils/heap.go (hunk 29 above)
Comment (soundvibe) [reply src/query/functions/utils/heap.go:38]: I'm not sure that having more code makes it more clear (especially in this case). I believe that logical `and` and `or` operators have their place and were added to programming languages for good reasons. I highly doubt that avoiding them makes code more clear.

File: src/query/functions/aggregation/take.go (hunk 33 above)
Comment (arnikola) [reply src/query/functions/aggregation/take.go:239]: +1

File: src/query/functions/utils/heap.go (hunk 29 above)
Comment (arnikola) [reply src/query/functions/utils/heap.go:38]: As a reviewer, it took a while to parse the logic as is; would recommend changing it to be more straightforward

File: src/query/functions/aggregation/take.go (hunk 27 above)
Comment (arnikola) [reply src/query/functions/aggregation/take.go:146]: Otherwise we end up with a ton of little utility functions that never really get reused since nobody else is aware of them

File: src/query/functions/linear/sort.go (hunk 32 above)
Comment (arnikola) [reply src/query/functions/linear/sort.go:117]: +1

File: src/query/functions/utils/heap.go (hunk 29 above)
Comment (soundvibe) [reply src/query/functions/utils/heap.go]: Simplified them and added some documentation to make it more clear.

File: src/query/functions/aggregation/take.go (hunk 33 above)
Comment (soundvibe) [reply src/query/functions/aggregation/take.go]: Here we do not care about indices so changed to `append`.

File: src/query/functions/linear/sort.go (hunk 32 above)
Comment (soundvibe) [reply src/query/functions/linear/sort.go:117]: We are relying on index when setting from `seriesMetas[i]` so `append` would just hide this and make it less clear.

File: src/query/functions/aggregation/take.go (hunk 27 above)
Comment (soundvibe) [reply src/query/functions/aggregation/take.go]: Removed `Min` function and used inline check.

File: src/query/functions/utils/heap.go [hunk 34]
```diff
@@ -33,17 +35,38 @@ type ValueIndexPair struct {
...
+	return i.Val > j.Val || lesserIfNaNs(i.Val, j.Val)
+}
+
+// Compares two floats for equality with NaNs taken into account.
+func equalWithNaNs(i,j float64) bool {
+	return i == j || math.IsNaN(i) && math.IsNaN(j)
+}
+
+// Checks which value is less if first of them is NaN.
+// Basically, we do not want to add NaNs to the heap when it has reached it's cap so this fn should be used to prevent this.
```
Comment (arnikola) [inline src/query/functions/utils/heap.go:57]: nit; `compareNaNs` maybe?

File: src/query/functions/linear/sort.go (hunk 32 above)
Comment (arnikola) [reply src/query/functions/linear/sort.go:117]: Yeah, it's just we've seen issues before with forward compat where future changes to slice size caused out of index errors, while append will deal with this case

**Review by arnikola — APPROVED** [review]
LGTM with a couple of small nits
